llvm/lib/Target/AIE/AIEPostPipeliner.cpp

```diff
@@ -0,0 +1,441 @@
+    for (auto &Dep : SU.Preds) {
+      const auto *Pred = Dep.getSUnit();
+      // Any predecessor in the first iteration
+      int K1 = Pred->NodeNum;
+      if (K1 < NInstr) {
+        const int Latest = Earliest - Dep.getSignedLatency();
+        Info[K1].Latest = std::min(Info[K1].Latest, Latest);
+      }
+    }
+  }
+  LLVM_DEBUG(for (int K = 0; K < NInstr; K++) {
+    dbgs() << "SU" << K << " : " << Info[K].Earliest << " - " << Info[K].Latest
+           << "\n";
+  });
+}
+
+void dumpGraph(int NInstr, const std::vector<NodeInfo> &Info,
+               ScheduleDAGInstrs *DAG) {
+  dbgs() << "digraph {\n";
+
+  for (int K = 0; K < NInstr; K++) {
+    auto &SU = DAG->SUnits[K];
+    for (auto &Dep : SU.Succs) {
+      auto *Succ = Dep.getSUnit();
+      int S = Succ->NodeNum;
+      if (S % NInstr == K) {
+        continue;
+      }
+
+      dbgs() << "\tSU" << K << " -> "
+             << "SU" << S;
+
+      if (S >= NInstr) {
+        dbgs() << "_" << S % NInstr;
+      }
+      dbgs() << "# L=" << Dep.getSignedLatency() << "\n";
+    }
+  }
+  dbgs() << "}\n";
+}
+
+int PostPipeliner::mostUrgent() {
+  assert(FirstUnscheduled < NInstr);
+  while (Info[FirstUnscheduled].Scheduled) {
+    FirstUnscheduled++;
+  }
+  assert(FirstUnscheduled < NInstr);
+
+  int Best = -1;
+  LLVM_DEBUG(dbgs() << "Available:");
+  for (int K = FirstUnscheduled; K < NInstr; K++) {
+    const auto &SU = DAG->SUnits[K];
+    // Check whether it is available
+    if (any_of(SU.Preds, [&](const auto &Dep) {
+          return !Info[Dep.getSUnit()->NodeNum].Scheduled;
+        })) {
+      continue;
+    }
+    LLVM_DEBUG(dbgs() << " SU" << K);
+    // Yeah, I know. This is a difficult way to schedule in the original
+    // node order. Have patience, my friend.
+    if (Best == -1) {
+      Best = K;
+      LLVM_DEBUG(dbgs() << "*");
+    }
+  }
+  LLVM_DEBUG(dbgs() << "\n");
+  assert(Best >= 0);
+  return Best;
+}
+
+bool PostPipeliner::scheduleFirstIteration() {
+  // Set up the basic schedule from the original instructions
+  for (int K = 0; K < NInstr; K++) {
+    const int N = mostUrgent();
+    LLVM_DEBUG(dbgs() << "  Trying " << N << "\n");
+    SUnit &SU = DAG->SUnits[N];
+    MachineInstr *MI = SU.getInstr();
+    const int Earliest = Info[N].Earliest;
+    // Find the first cycle that fits. We try every position modulo II
+    const int Actual = fit(MI, Earliest, II, II);
+    if (Actual < 0) {
+      // out of resources for this II;
+      LLVM_DEBUG(dbgs() << "Out of resources\n");
+      return false;
+    }
+    const int LocalCycle = Actual % II;
+    const MemoryBankBits MemoryBanks = HR.getMemoryBanks(MI);
+    LLVM_DEBUG(dbgs() << "  Emit in " << -Depth + LocalCycle << "\n");
+    int Cycle = -Depth + LocalCycle;
+    LLVM_DEBUG(dbgs() << "  Emit in " << Cycle << "\n");
+    HR.emitInScoreboard(Scoreboard, MI->getDesc(), MemoryBanks, MI->operands(),
+                        MI->getMF()->getRegInfo(), Cycle);
```

Comment:
I'm not debating for correctness, but I wonder if we should not immediately emit the resources at C+II, C+2II, etc. Otherwise, we will only see conflicts due to resources that "wrap around the II" when scheduling the next iterations. At that point, it's too late and we will try a larger II. Maybe if we could anticipate those resources, we could find a better linear schedule that has less conflicts when scheduling all copies.

Reply:
I agree, but that can't be done with the current HazardRecognizer/Scoreboard, since I would need to push the modulo limit into the scoreboard handling. I will create a follow up to do that.

Reply:
I mean, you know the II and NCopies, you could just have a `for` loop here instead of a single call to `HR.emitInScoreboard`? (Not in this PR, agreed)

Reply:
Ah well, it could be done by having enough depth in the scoreboard. Anyway, given that it doesn't impact correctness, I would push that to a follow up.